bqskit/compiler/passes/partitioning/scan.py

```diff
@@ -180,59 +265,41 @@ def run(self, circuit: Circuit, data: dict[str, Any]) -> None:
             for qudit, amount in enumerate(amount_to_add_to_each_qudit):
                 divider[qudit] += amount
 
-            # Find the scores of the qudit groups.
-            best_score = None
-            best_region = None
-
-            for qudit_group in qudit_groups:
-
-                ops_and_cycles = circuit.operations_with_cycles(
-                    qudits_or_region=CircuitRegion({
-                        qudit_index: (divider[qudit_index], num_cycles)
-                        for qudit_index in qudit_group
-                    }),
-                )
-
-                in_qudits = list(q for q in qudit_group)
-                stopped_cycles = {q: num_cycles for q in qudit_group}
-                score = 0
-
-                for cycle, op in ops_and_cycles:
-                    if len(op.location.union(in_qudits)) != len(in_qudits):
-                        for qudit_index in op.location.intersection(in_qudits):
-                            stopped_cycles[qudit_index] = cycle
-                            in_qudits.remove(qudit_index)
-                    else:
-                        if len(op.location) > 1:
-                            score += self.multi_gate_score
-                        else:
-                            score += self.single_gate_score
-
-                    if len(in_qudits) == 0:
-                        break
-
-                if best_score is None or score > best_score:
-                    best_score = score
-                    best_region = CircuitRegion({
-                        qudit: (
-                            divider[qudit],
-                            # Might have errors if below is removed
-                            stopped_cycles[qudit] - 1,
-                        )
-                        for qudit in qudit_group
-                        # This statement
-                        # if stopped_cycles[qudit] - 1 >= divider[qudit]
-                    })
+            # Heap partitioning
+            # Get best (up to date) block
+            while len(score_heap) > 0:
+                (best_score, group_index) = heapq.heappop(score_heap)
+                best_score *= -1
+                if best_score == block_list[group_index].score:
```

Comment:
Why is this check necessary? Can't you just check if the heapq is not empty and pop the top one?

Reply:
Because I don't clear of out of date scores during the rescoring. This makes sure we're not using an out of date score to select the next best region.

Reply:
Is not clearing them worth it? Clearing them could be done in logarithm time since the array is sorted and we know the scores of the old ones. Not clearing them, leads to increasing heap size. Every time a block is formed, we have to add probably hundreds of elements to the heap. Imagine a large circuit with tens of thousands to millions of blocks, this can easily lead to an out-of-memory error.

Reply:
I think clearing the old scores would still take linear time (because it's just a heap) but you're probably right about clearing it. Do you think it might be worth changing to a BST? Getting the max score would take O(log n) instead of O(1), but cleaning the old scores out would take O(k log n) instead of O(k n).

Reply:
If you remember the old scores, you can find them in O(log n) time in the heap. Removing an element in the heap also takes O(log n) time. This process can be repeated for each of the k elements you would like to remove, so O(k log n).